src/framework/components/joint/component.js

```diff
@@ -0,0 +1,371 @@
+        rot.setFromMat4(pcTransform);
+
+        const ammoVec = new Ammo.btVector3(pos.x, pos.y, pos.z);
+        const ammoQuat = new Ammo.btQuaternion(rot.x, rot.y, rot.z, rot.w);
+
+        ammoTransform.setOrigin(ammoVec);
+        ammoTransform.setRotation(ammoQuat);
+
+        Ammo.destroy(ammoVec);
+        Ammo.destroy(ammoQuat);
+    }
+
+    _updateAngularLimits() {
+        const constraint = this._constraint;
+        if (constraint) {
+            const alx = this._angularLimitsX;
+            const aly = this._angularLimitsY;
+            const alz = this._angularLimitsZ;
+
+            const limits = new Ammo.btVector3(alx.x, aly.x, alz.x);
+            constraint.setAngularLowerLimit(limits);
+            limits.setValue(alx.y, aly.y, alz.y);
+            constraint.setAngularUpperLimit(limits);
+            Ammo.destroy(limits);
+        }
+    }
+
+    _updateLinearLimits() {
+        const constraint = this._constraint;
+        if (constraint) {
+            const llx = this._linearLimitsX;
+            const lly = this._linearLimitsY;
+            const llz = this._linearLimitsZ;
+
+            const limits = new Ammo.btVector3(llx.x, lly.x, llz.x);
+            constraint.setLinearLowerLimit(limits);
+            limits.setValue(llx.y, lly.y, llz.y);
+            constraint.setLinearUpperLimit(limits);
+            Ammo.destroy(limits);
+        }
+    }
+
+    _createConstraint() {
+        if (this._entityA && this._entityA.rigidbody) {
+            this._destroyConstraint();
+
+            const mat = new Mat4();
+
+            const bodyA = this._entityA.rigidbody.body;
+            bodyA.activate();
+
+            const jointWtm = this.entity.getWorldTransform();
+
+            const entityAWtm = this._entityA.getWorldTransform();
+            const invEntityAWtm = entityAWtm.clone().invert();
+            mat.mul2(invEntityAWtm, jointWtm);
+
+            const frameA = new Ammo.btTransform();
+            this._convertTransform(mat, frameA);
+
+            if (this._entityB && this._entityB.rigidbody) {
+                const bodyB = this._entityB.rigidbody.body;
+                bodyB.activate();
+
+                const entityBWtm = this._entityB.getWorldTransform();
+                const invEntityBWtm = entityBWtm.clone().invert();
+                mat.mul2(invEntityBWtm, jointWtm);
+
+                const frameB = new Ammo.btTransform();
+                this._convertTransform(mat, frameB);
+
+                this._constraint = new Ammo.btGeneric6DofSpringConstraint(bodyA, bodyB, frameA, frameB, !this._enableCollision);
+
+                Ammo.destroy(frameB);
+            } else {
+                this._constraint = new Ammo.btGeneric6DofSpringConstraint(bodyA, frameA, !this._enableCollision);
+            }
+
+            Ammo.destroy(frameA);
+
+            const axis = ['X', 'Y', 'Z', 'X', 'Y', 'Z'];
+
+            for (let i = 0; i < 6; i++) {
+                const type = i < 3 ? 'linear' : 'angular';
```

Comment:
change to '_linear' and '_angular' to avoid adding '_" later

Reply:
Done.